Fixes to CILogon / `openid_connect` Tests

aaronskiba opened the pull request:
Fixes #921

Changes proposed in this PR:
- Fix flaky tests due to overriding of `OmniAuth.config.mock_auth[:openid_connect]` [648d6d8](https://github.com/portagenetwork/roadmap/pull/922/commits/648d6d848832b31ddbca4fa7420070e4aaef219b)
  - Prior to this PR, whenever `OmniAuth.config.mock_auth[:openid_connect]` was being overridden in this file, the value was always subsequently assigned to `@request.env['omniauth.auth']`. These changes simply bypass `OmniAuth.config.mock_auth[:openid_connect]` and directly assign `OmniAuth::AuthHash.new(...)` to `@request.env['omniauth.auth']` instead.

- Update and re-enable openid_connect features tests [4fd0f55](https://github.com/portagenetwork/roadmap/pull/922/commits/4fd0f55d5a0faa73ab979aef7f58b709f84659c6)
  - config/secrets.yml
    - `Rails.application.secrets.omniauth_full_host` does not appear to be necessary in test mode. In fact, if a value exists for `ENV['OMNIAUTH_FULL_HOST']` (e.g. the value set for development) then it results in the tests failing (By default, `Capybara.server_port` is set to a random value, so hardcoding a secret isn't completely straightforward.)
  - spec/integration/openid_connect_sso_spec.rb
    - Here we are re-enabling the "links account from external credentails" test, which was failing due to the aforementioned issue with `config/secrets.yml`.
    - We are also adding `:js`, which if desired, makes it simpler to execute these tests in non-headless mode.

aaronskiba committed:
Address  `OmniAuth.config.mock_auth` overriding

This commit is meant to resolve the following issue that resulted from the overriding of `OmniAuth.config.mock_auth[:openid_connect]`: https://github.com/portagenetwork/roadmap/issues/921

Whenever `OmniAuth.config.mock_auth[:openid_connect]` was being overridden in this file,  the value was always subsequently assigned to `@request.env['omniauth.auth']`. These changes simply bypass `OmniAuth.config.mock_auth[:openid_connect]` and directly assign `OmniAuth::AuthHash.new(...)` to `@request.env['omniauth.auth']` instead.

aaronskiba committed:
Update and re-enable openid_connect features tests

Update and re-enable openid_connect features tests

config/secrets.yml
- `Rails.application.secrets.omniauth_full_host` does not appear to be necessary in test mode. In fact, if a  value exists for `ENV['OMNIAUTH_FULL_HOST']` (e.g. the value set for development) then it results in the tests failing (By default, `Capybara.server_port` is set to a random value, so hardcoding a secret isn't completely straightforward.)

spec/integration/openid_connect_sso_spec.rb
- Here we are re-enabling the "links account from external credentails" test, which was failing due to the aforementioned issue with `config/secrets.yml`.
- We are also adding `:js`, which if desired, makes it simpler to execute these tests in non-headless mode.

lagoan commented on spec/integration/openid_connect_sso_spec.rb:
```diff
@@ -24,7 +24,7 @@
     end
 
-    it 'creates account from external credentials' do
+    it 'creates account from external credentials', :js do
```
These tests continue to be flaky

lagoan commented on spec/integration/openid_connect_sso_spec.rb:
```diff
@@ -51,7 +51,7 @@
     end
 
-    xit 'links account from external credentails' do
+    it 'links account from external credentails', :js do
```
These tests continue to be flaky

aaronskiba replied:
You are right, thanks for catching that.
`rspec ./spec/controllers/omniauth_callbacks_controller_spec.rb ./spec/integration --seed=10840` is at least one way to trigger failed tests.

lagoan commented on spec/controllers/omniauth_callbacks_controller_spec.rb:
```diff
@@ -15,19 +15,17 @@
-
     # Assign the mocked authentication hash to the request environment
-    @request.env['omniauth.auth'] = OmniAuth.config.mock_auth[:openid_connect]
+    @request.env['omniauth.auth'] = OmniAuth::AuthHash.new({
+                                                             provider: 'openid_connect',
+                                                             uid: '12345',
+                                                             info: {
+                                                               email: 'user@organization.ca',
+                                                               first_name: 'Test',
+                                                               last_name: 'User',
+                                                               name: 'Test User'
+                                                             }
+                                                           })
```
Its been a while since I worked on this but believe we dont need to define ``@request.env['omniauth.auth'] ``

aaronskiba replied:
> Its been a while since I worked on this but believe we dont need to define `@request.env['omniauth.auth'] `

I'm thinking we might need it. The `def openid_connect` method (`app/controllers/users/omniauth_callbacks_controller.rb`) has `auth = request.env['omniauth.auth']`. The tests appear to be failing without it: https://github.com/portagenetwork/roadmap/actions/runs/11409377997/job/31749474766

Rather than using `@request.env['omniauth.auth'] = OmniAuth::AuthHash.new({...` here, maybe we could  use `Rails.application.env_config['omniauth.auth'] = OmniAuth.config.mock_auth[:openid_connect]` like we did in `spec/integration/openid_connect_sso_spec.rb`?

aaronskiba replied:
Also, I was conflating `@request.env` with `Rails.application.env_config`. Maybe there is a conflict occurring between these two.

aaronskiba replied:
I think the real issue was that the overriding of `User.from_omniauth` in `spec/controllers/omniauth_callbacks_controller_spec.rb` was persisting beyond that test file. I've addressed the issue here: https://github.com/portagenetwork/roadmap/pull/922/commits/8d2fd866ab6245e6311a6bdf941ab1cf7043c3c8

lagoan reviewed (changes requested):
There are a couple of fixes here that I think can be improved. I left comments on this PR and you can check this [branch](https://github.com/portagenetwork/roadmap/tree/omar/issues/921)  for reference on some ideas.

aaronskiba committed:
Address `User.from_omniauth` overriding

- The code removed via this commit included the comment `# Reset the `from_omniauth` method after each test`. However, rather than restoring the method, the subsequent code redefined it.
- As a result, when executed in test files run after this, `User.method(:from_omniauth).source_location` would evaluate to `/spec/controllers/omniauth_callbacks_controller_spec.rb", 109]`.
- This commit restores `from_omniauth` to its actual User class definition in the `after do` block of `omniauth_callbacks_controller_spec.rb`.

aaronskiba committed:
Refactor: Create `:openid_connect` trait

aaronskiba committed:
Refactor/Cleanup

- `OmniAuth.config.mock_auth[:openid_connect]` is defined in `spec/spec_helper.rb`. These changes remove the custom hash and instead use that pre-made one.

lagoan reviewed (changes requested):
We just need a CHANGELOG entry and we are set.

aaronskiba committed:
Update CHANGELOG.md

lagoan reviewed (approved):
LGTM!
